src/lib/Tisch/Internal/Create.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedLabels      #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE QuasiQuotes           #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+-- | This module deals with generating database creation statements from a
+-- Tisch database description.
+module Tisch.Internal.Create
+ ( makeCreateStatement
+ ) where
+
+import           Data.Proxy
+import           Data.Singletons
+import           Data.Text                     as T
+import           GHC.TypeLits
+import qualified Opaleye.PGTypes               as O
+import           Text.InterpolatedString.Perl6
+import           Tisch
+import           Tisch.Internal.Table
+
+-- | Generate a CREATE statement for a given table
+makeCreateStatement
+  :: forall table
+   . ( Tisch.Internal.Table.All ColumnInfo (Columns table)
+     , ConsumeTypeList (Columns table)
+     , KnownSymbol (TableName table)
+     )
+  => Table table
+  -> Text
+makeCreateStatement = colInfosCreateStatement (pack tableName) . colInfos
+  where
+    tableName = symbolVal (Proxy @(TableName table))
+
+-- | Get the term-level information about a table
+colInfos
+  :: forall table
+   . ( Tisch.Internal.Table.All ColumnInfo (Columns table)
+     , ConsumeTypeList (Columns table)
+     )
+  => Table table
+  -> [ColInfo]
+colInfos _ = dis (Proxy @ColumnInfo) colInfo (Proxy @(Columns table))
+
+-- TODO: Support default and read only
+-- | Get a create statement from a description of a table.
+colInfosCreateStatement :: Text -> [ColInfo] -> Text
+colInfosCreateStatement tableName is
+  = [qc|CREATE TABLE "{tableName}" ({intercalate "," args});|]
+  where
+    args = colInfoCreateArg <$> is
+    colInfoCreateArg :: ColInfo -> Text
+    colInfoCreateArg (Column name _wcap rcap unique pgtype ()) =
+      let uniqueSuffix :: Text
+          uniqueSuffix = case unique of
+                           NotUnique -> ""
+                           Unique    -> " UNIQUE"
+          nnSuffix :: Text
+          nnSuffix = case rcap of
+                       R  -> " NOT NULL"
+                       RN -> ""
+      in [qc|"{name}" {unPGType pgtype}{uniqueSuffix}{nnSuffix}|]
```

Comment:
The same thing that makes `TemplateHaskell` bad: The stage restrictions that come with it. But also, and this is on a more subjective note: They bring new syntax which I don't really care for, and it is not something that is so hard to achieve by other means that you do need TH or QQ. This is just a small detail, and it could be easily written using, say, `mconcat`.